mqtt/mqtt-broker-core/src/auth/authentication.rs

```diff
@@ -43,34 +34,89 @@ pub trait Authenticator {
     /// * `Err(e)` - an error occurred when authenticating a client.
     async fn authenticate(
         &self,
-        username: Option<String>,
-        credentials: Credentials,
+        context: AuthenticationContext,
     ) -> Result<Option<AuthId>, Self::Error>;
 }
 
+/// A data required to authenticate connected client.
+#[derive(Debug)]
+pub struct AuthenticationContext {
+    client_id: ClientId,
+    peer_addr: SocketAddr,
+    username: Option<String>,
+    password: Option<String>,
+    certificate: Option<Certificate>,
+}
+
+impl AuthenticationContext {
+    pub fn new(client_id: ClientId, peer_addr: SocketAddr) -> Self {
+        Self {
+            client_id,
+            peer_addr,
+            username: None,
+            password: None,
+            certificate: None,
+        }
+    }
+
+    pub fn with_username(&mut self, username: impl Into<String>) -> &mut Self {
```

Comment:
q: nit: so for builder pattern should we use "with_xxxx" everywhere?

Reply:
this specific case is different: it an object that optional fields set in "builder-like" style. usually the builder doesn't expose getters.

But, seems like a good idea to have similar semantics across all the code